lib/customisation.nix

```diff
@@ -102,8 +125,64 @@ rec {
         lib.setFunctionArgs result (lib.functionArgs result) // {
           override = overrideArgs;
         }
-      else result)
-    (lib.functionArgs f);
+        else result);
+
+  /* Modify an overridable and its override results.
+
+     An overridable is an attribute set with attributes providing
+     [overriding](#chap-overrides) functionality.
+     These attributes contains the information (state) of the function that
+     produces the overridable, and the argument passed to such function.
+
+     Such encapsulated state doesn't change along with the modified
+     attribute set, and the modification needs to be re-applied to each
+     overriding result. `lib.applyToOverridable` automates the procedure by
+     decorating those attributes.
+
+     `override`, `overrideAttributes`, `overrideDerivation` and `expand`
+     are chosen to cover most of the use cases.
+
+     Example:
+       helloWithAns1 = lib.expandDerivation true { ans = 42; } pkgs.hello
+
+       helloWithAns1.ans
+       => 42
+       (helloWithAns1.override { }).ans
+       => error: attribute 'ans' missing
+
+       helloWithAns2 = lib.applyToOverridable (lib.expandDerivation true { ans = 42; }) pkgs.hello
+
+       helloWithAns2.ans
+       => 42
+       (helloWithAns2.override { }).ans
+       => 42
+
+     Type:
+       applyToOverridable :: (AttrSet -> AttrSet) -> AttrSet -> AttrSet
+  */
+  applyToOverridable = modify: overridable:
+    let
+      r = modify overridable;
+    in
+    r // {
+      ${if r ? override then "override" else null} = lib.copyFunctionArgs r.override (fdrv: applyToOverridable modify (r.override fdrv));
+      ${if r ? overrideAttrs then "overrideAttrs" else null} = fdrv: applyToOverridable modify (r.overrideAttrs fdrv);
+      ${if r ? overrideDerivation then "overrideDerivation" else null} = fdrv: applyToOverridable modify (r.overrideDerivation fdrv);
```

Comment:
`overrideDerivation` is not an attribute, instead it's a function, `lib.overrideDerivation`.

Reply:
> `overrideDerivation` is not an attribute, instead it's a function, `lib.overrideDerivation`.

But [`<pkg>.overrideDerivation`](https://nixos.org/manual/nixpkgs/unstable/#sec-pkg-overrideDerivation) does exist. This attribute seems to be added by `makeOverridable`.

Reply:
Ah you're right! Never mind :)